TotalCrossSDK/src/main/java/totalcross/json/JSONFactory.java

```diff
@@ -110,95 +119,112 @@ public void setDescription(String description)
     }
     return parse(outerObject, jsonArray, classOfT);
   }
-  
-  public static <T> T parse(JSONObject jsonObject, Class<T> classOfT) throws InstantiationException,
-  IllegalAccessException, IllegalArgumentException, InvocationTargetException, JSONException, NoSuchMethodException, SecurityException {
-      return parse(null, jsonObject, classOfT);
+
+  public static <T> T parse(JSONObject jsonObject, Class<T> classOfT)
+      throws InstantiationException, IllegalAccessException, IllegalArgumentException, InvocationTargetException,
+      JSONException, NoSuchMethodException, SecurityException {
+    return parse(null, jsonObject, classOfT);
   }
 
-  private static <T> T parse(Object outerObject, JSONObject jsonObject, Class<T> classOfT) throws InstantiationException,
-      IllegalAccessException, IllegalArgumentException, InvocationTargetException, JSONException, NoSuchMethodException, SecurityException {
+  private static <T> T parse(Object outerObject, JSONObject jsonObject, Class<T> classOfT)
+      throws InstantiationException, IllegalAccessException, IllegalArgumentException, InvocationTargetException,
+      JSONException, NoSuchMethodException, SecurityException {
     if (classOfT.isArray()) {
       throw new IllegalArgumentException();
     }
     T object = null;
     try {
-        object = classOfT.newInstance();
+      object = classOfT.newInstance();
```

Comment:
It seems to me that the creation of a single object based on its class could also be cached, in a similar way that it is possible to cache methods.

It should be easier to read if the creation of the object was simply:

```java
T object = createNewInstanceOfT(classOfT, outerObject);
```

and let `createNewInstanceOfT` throws all the exceptions. `createNewInstanceOfT` could be like this (not taking care of race conditions yet):

```java
private static final Supplier<Object> FORBIDEN = () -> {
  throw new RuntimeException("oops");
};

private static final Function<Object, Object> FORBIDEN_FUNCTION = () -> {
  throw new RuntimeException("oops");
};

private static <T> T createNewInstanceOfT(Class<T> classOfT, Object outerObject) {
  Supplier<T> supplierOfT = (Supplier<T>) cacheCtor.get(claffOfT);
  if (supplierOfT != null && supplierOfT != FORBIDEN) {
    return supplierOfT .get();
  } else if (supplierOfT == null) {
    try {
      T obj = classOfT.newInstance();
      cacheCtor,put(classOfT, classOfT::newInstance); // working around exceptions, not literally RACE-CONDITION-AWARE
      return obj;
    } catch (Exception e) {
      cacheCtor.put(classOfT, FORBIDEN); // mark such that won't try to fetch default ctor again RACE-CONDITION-AWARE
    }
  }
  if (outerObject == null ||  classOfT.getName().indexOf(outerObject.getClass().getName()) == -1) {
    // CANNOT INSTANTIATE
    throw new ProperCantInstantiateException();
  }
  HashMap<Class<?>, Function<Object, T>> mapConstructorOfTWithParam = (Function<Object, T>) cacheCtorWithParam(classOfT);
  if (mapConstructorOfTWithParam != null) {
    Function<Object, T> constructorOfTWithParam = mapConstructorOfTWithParam .get(outerObject.getClass());
    if (constructorOfTWithParam  == FORBIDEN_FUNCTION ) {
      // CANNOT INSTANTIATE
      throw new ProperCantInstantiateException();
    } else if (constructorOfTWithParam != null) {
      return constructorOfTWithParam.apply(outerObject);
    }
  } else {
    mapConstructorOfTWithParam = new HashMap<>();
    cacheCtorWithParam.put(classOfT, mapConstructorOfTWithParam); // RACE-CONDITION-AWARE
  }
  Constructor<T> constructorOfT = classOfT.getDeclaredConstructor(outerObject.getClass());
  if (constructorOfT == null) {
    mapConstructorOfTWithParam.put(outerObject.getClass(), FORBIDEN_FUNCTION); // mark such that won't try to fetch ctor with this class again RACE-CONDITION-AWARE
    // CANNOT INSTANTIATE
    throw new ProperCantInstantiateException();
  }
  mapConstructorOfTWithParam.put(outerObject.getClass(), oObject -> constructorOfT.newInstance(oObject)); // working around exceptions, not literally RACE-CONDITION-AWARE
  return constructorOfT.newInstance(outerObject);
}
```

> I took some extra care with situations where it won't be possible to create such objects, but perharps this shouldn't?
> The idea was to get a proper headstart to know _which_ ctor to call, but if it is invalid then _MAYBE_ my worries are just overengineering